### ISSUE TITLE
[REVIEW]: UPOsHam:  A Python package for computing unstable periodic orbits in two degrees of freedom Hamiltonian systems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1684]
**Submitting author:** @WyLyu (<a href="http://orcid.org/0000-0003-2570-9879">WenYang Lyu</a>)
**Repository:** <a href="https://github.com/WyLyu/UPOsHam" target ="_blank">https://github.com/WyLyu/UPOsHam</a>
**Version:** v0.1.0
**Editor:** @kthyng
**Reviewer:** @dpsanders, @JoshKarpel
**Archive:** <a href="https://doi.org/10.5281/zenodo.3606676" target="_blank">10.5281/zenodo.3606676</a>

## Status

[![status](https://joss.theoj.org/papers/bc6d9d9b9254ef94543016a4c40fafd0/status.svg)](https://joss.theoj.org/papers/bc6d9d9b9254ef94543016a4c40fafd0)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/bc6d9d9b9254ef94543016a4c40fafd0"><img src="https://joss.theoj.org/papers/bc6d9d9b9254ef94543016a4c40fafd0/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/bc6d9d9b9254ef94543016a4c40fafd0/status.svg)](https://joss.theoj.org/papers/bc6d9d9b9254ef94543016a4c40fafd0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/WyLyu/UPOsHam" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@dpsanders & @JoshKarpel, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @kthyng know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/WyLyu/UPOsHam">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@WyLyu) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/WyLyu/UPOsHam">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@WyLyu) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by kthyng** [comment]
@JoshKarpel Thank you for your question and work on your review, I will consult with the editorial board and get back to you soon.

**Comment by Shibabrat** [comment]
@kthyng I'm willing to explain the design decision of this package because it is related to the theory and applications of this method. Should I create a separate issue or briefly describe it here? Not sure about the etiquette at this point of review period.

Thanks

**Comment by kthyng** [comment]
@Shibabrat Why don't you give a brief description of the design decision here just to help deciding how to move forward, for now. One design suggestion is for you to change this to a library so that a user doesn't have to copy and paste to use it.

**Comment by Shibabrat** [comment]
Ok. 

We have seen that researchers working in Hamiltonian dynamics have their favorite method to compute unstable periodic orbits. These methods are popular in chemical reactions, celestial mechanics, and engineering dynamics and always end up in appendices of the papers. Some of these papers are also in our references. 

However, for anyone new or outside of the aforementioned communities, it becomes a slow tedious process to go through the code (even if available) and to adapt it for a specific problem. We think an illustration style set of transparent scripts that has minimal dependencies for installation, is easy to change and modify can help with learning the underlying methods. 

Next, examples in the package are also in the form that is being used for ongoing work in our group on chemical reaction dynamics and hence gives a direct way to reproduce the results.

**Comment by kthyng** [comment]
@Shibabrat Your argument that this software is useful because it demonstrates going through some physics problems is convincing. However, it would be much better if your software was a proper python package that can be used in a typical manner in order to meet your goals. Specifically, can you change your setup so that another researcher using it can avoid copying and pasting bits of code into a new project, and instead import the full package and use reusable functions?

@JoshKarpel We think that the submission can be substantially improved, as per my first comment here, though your concern with the one-off modifications is ok — the point of the requirement is that the software have reuse capacity, which this appears to have. Still, changing this to a package will also help to address your concern.

**Comment by JoshKarpel** [comment]
@kthyng Sounds good to me. I think if it is not possible to make it more package-y, it needs to be very, very clear how and where to copy-paste-modify. But I would prefer a more package-y solution.

@Shibabrat  @WyLyu some semi-concrete thoughts on how to move toward a package-based design...

So, my understanding is that there are three solution methods, which you use to solve three problems, leading to nine examples. Is it possible to decouple the code for each solution method from the problem it is solving, such that the same solution method code could be applied to any of the problems with only top-level configuration? As an incidental benefit, this would reduce the amount of code duplication in the project...

As a very concrete example, think about how the differential equation solvers in SciPy work ( https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp). There's a high-level solver which can apply different methods (set via the `method` argument) to whatever the user passes in. Many different differential equations in many different forms can be solved by the solver (although the user may need to select the correct method for their given problem, perhaps equivalent to selecting a solution method in your code?).

**Comment by Shibabrat** [comment]
@JoshKarpel Yes, you are correct. Some (I point to those below) details of the methods can not be decoupled from the specific features of the problem. In particular, one such feature is a detailed knowledge of the bottleneck region of a potential energy surface (PES). For a given molecular system, the PES is the output of a quantum chemistry calculation and the results are highly dependent on the molecule.

The first-order correction based on the monodromy matrix for the differential correction needs derivation based on how the bottleneck is oriented in the configuration space. The correction term is different in https://github.com/WyLyu/UPOsHam/blob/master/diffcorr_UPOsHam2dof.py#L744
from https://github.com/WyLyu/UPOsHam/blob/master/diffcorr_UPOsHam2dof.py#L759, and it can also be same as in https://github.com/WyLyu/UPOsHam/blob/master/diffcorr_UPOsHam2dof.py#L752. This correction is applied to the phase space coordinate that is most natural for the bottleneck of the PES. Fixing this coordinate also determines the coordinate that is used to check convergence (https://github.com/WyLyu/UPOsHam/blob/master/diffcorr_UPOsHam2dof.py#L651) and the one that is used to terminate the integration using event function (https://github.com/WyLyu/UPOsHam/blob/master/diffcorr_UPOsHam2dof.py#L303).

The turning point based on coordinate difference relies on the orientation of the PES to choose whether to use x or y or both coordinates to check which direction the trajectory is turning. In this case, the orientation of the bottleneck of the PES determines how the turning direction is checked (https://github.com/WyLyu/UPOsHam/blob/master/tpcd_UPOsHam2dof.py#L362).

The turning point method probably can be generalized but then one needs to manually choose two initial conditions near the final/true UPO. This is probably the one that comes closest to being written without problem-specific details as long as the user starts with two assumptions: first, the two starting initial conditions for the method should be close (whatever that means for the problem) to the final UPO and second the accuracy can't be improved beyond 10e-3 (which is fine for rough estimates or to restart the TPCD method but not for using the UPO in other quantitative calculations such as transition fraction or reaction flux)

Having said that, I do think some portions of the package can be separated as user-supplied functions with specific signatures, as per your suggestion.

**Comment by JoshKarpel** [comment]
@Shibabrat  When I see a block like https://github.com/WyLyu/UPOsHam/blob/master/diffcorr_UPOsHam2dof.py#L738 , it seems clear to me that this calculation is part of specifying the problem to be solved, and should be passed into the solver by the user. In that case it looks like the input parameters are `par`, `x0`, `x1`, and `phi_t1`, and the function should return the corrected `x0`.

If that means you have to pass ten different functions (or parameters, or whatever) into the solver, so be it: **in the current version the user would have to do that anyway, but by copy-paste-editing long functions instead of writing their own small, focused functions**. I think that passing configuration to the solver, even if there is a lot of it, is a far friendlier than being asked to make specific edits in hundreds of lines of code.

The examples would then largely be the definitions of the functions and configuration parameters for the appropriate problems, which would be much easier for people to take and re-purpose for their own problems.

**Comment by Shibabrat** [comment]
We have updated the methods by separating the problem-specific functions. The demonstration script which loads the method as module is the starting point for a user to modify according to their problem of interest.

**Comment by JoshKarpel** [comment]
I poked around in the latest version, and I definitely like where this is heading! Now that I think we're on the right architectural path for a reusable package, I'll start making issues in the repository again to address specific issues.

**Comment by kthyng** [comment]
Hi @dpsanders Will you be able to work on your review very soon?

**Comment by dpsanders** [comment]
@kthyng Sorry, I hope to be able to during the next few days.

**Comment by dpsanders** [comment]
I suggest that all future changes to the repo are done via Pull Requests, so that we can easily see what has been changed.

**Comment by dpsanders** [comment]
(Each Pull Request should ideally address a single, or a few related, points from the reviewers.)

**Comment by dpsanders** [comment]
The suggestions made by @JoshKarpel about the code structure are spot on. There is still a lot of duplication between files like `run_diffcorr_POfam_coupled.py` and `run_turningpoint_POfam_coupled.py` that start out by defining exactly the same functions (as far as I can tell).

This makes it difficult to know if any copy-paste errors have been introduced and difficult to identify the basic structure and key parts of the code.

All of this code duplication should be eliminated.

**Comment by dpsanders** [comment]
There are also still no automated tests.

**Comment by JoshKarpel** [comment]
@Shibabrat  @WyLyu please make sure to reference the issues (and specific commits, where reasonable) from the pull requests so we know which is addressing what.

**Comment by Shibabrat** [comment]
We've mostly done the last few changes that include the filename, moving code to different directories, and structuring documentation using pull requests. But maybe we didn't tag or reference the issues so that's why it's not clear what we have been changing. Apologies for the confusion. We are cutting our teeth with the branching and pull requests on different but related issues.

In summary, we are working on the automated tests and removing the duplication in the scripts. These sets of changes will also address how to run the scripts without the figures interfering with the progress of the iteration.

**Comment by JoshKarpel** [comment]
@Shibabrat  @WyLyu The PR churn is a little hard to follow. I'd appreciate it if you could batch your changes a little more, explain what's going on in each PR, and write comments in the PRs explaining which issues you're addressing with them. Not everything needs to be merged to master immediately; you can checkout a branch that someone else pushed to GitHub and work on it locally, then push more commits to it, without merging the PR.

**Comment by kthyng** [comment]
@Shibabrat @WyLyu What is your status on comments from your reviewers?

**Comment by Shibabrat** [comment]
We have now started two PRs (https://github.com/WyLyu/UPOsHam/pull/34, https://github.com/WyLyu/UPOsHam/pull/35) that will resolve most of the issues currently pending. @JoshKarpel and @dpsanders can you please let us know if it makes sense what we have addressed? We can then go ahead and merge them or make appropriate changes.

**Comment by kthyng** [comment]
Just another quick ping to check in on thoughts from @JoshKarpel and @dpsanders!

**Comment by JoshKarpel** [comment]
@kthyng We're definitely making progress. The authors have made a lot of good changes, and I think this big PR finally gets us to a place where they can start satisfying the review requirements directly (things like setting up documentation and making it an installable package).

**Comment by kthyng** [comment]
Awesome!

**Comment by labarba** [comment]
👋  @dpsanders — Your last comment from Oct. 17 says:
> There are also still no automated tests.

Do note that JOSS review criteria regarding [tests](https://joss.readthedocs.io/en/latest/review_criteria.html#tests) is that "Authors are strongly encouraged to include an automated test suite…" but "Documented manual steps that can be followed to objectively check the expected functionality of the software" are OK.

Given this, and that this review has been ongoing for three months already, and the authors have made many improvements, are you ready to submit a recommendation?

**Comment by dpsanders** [comment]
There has definitely been a lot of progress, but as @JoshKarpel comments above, there are still some fundamental issues that the authors have not fully addressed, such as docs.

Also, I do not see any tests where something

**Comment by Shibabrat** [comment]
thank you @labarba 

@dpsanders and @JoshKarpel the package documentation is under progress (https://uposham.readthedocs.io/en/latest/#) since the recent PR contains the updated docstrings and changes to calling the scripts from the command line.

Test (comparing the analytical and numerical solution) in the form of a notebook will also reflect these changes.

**Comment by Shibabrat** [comment]
@kthyng @dpsanders and @JoshKarpel we have made some recent updates to the documentation and tests (this is available in the test directory as tests.ipynb) that are compatible with the module structure of the methods. 

The correctness test will soon be restructured as an assert test along with the visual comparison (at the moment).

API documentation will feature example code snippets and reference to a script that solves the provided example systems using a method. 

We would appreciate some feedback on these new developments.

**Comment by JoshKarpel** [comment]
I submitted some new issues with high-level feedback on the docs.

The tests are moving in the right direction; very happy to hear that you'll be able to do an actual assertion. There should be some information in the README about how to run the tests and how to do any manual verification, but that can wait until the structure settles down.

**Comment by kthyng** [comment]
@Shibabrat Looks like your documentation should be your main focus now.

**Comment by Shibabrat** [comment]
@dpsanders @JoshKarpel Can you take a look at the following PRs?

We've updated the documentation and started a PR: https://github.com/WyLyu/UPOsHam/pull/50. This also includes a plain text explanation of how the methods and the systems fit together. 

The test branch and it's PR: https://github.com/WyLyu/UPOsHam/pull/49. This includes the unit test for all the methods.

Both these should address most of the pending issues and requirements upon merging. We look forward to your feedback.

**Comment by Shibabrat** [comment]
@kthyng We have now updated the documentation (https://uposham.readthedocs.io/en/latest/) and the new tests have also passed (https://github.com/WyLyu/UPOsHam/pull/49). How do we proceed now?

**Comment by kthyng** [comment]
@Shibabrat Looks like [#49](https://github.com/WyLyu/UPOsHam/pull/49) is good to go from the reviewers and maybe we can hear on [#50](https://github.com/WyLyu/UPOsHam/pull/50) from @dpsanders. 

I see there are a few more small tasks to clean up.  Otherwise, @dpsanders and @JoshKarpel, could you give an overview of what's left to do at this point? I see unchecked boxes but perhaps some can be checked off now? Thanks very much.

**Comment by JoshKarpel** [comment]
The last major fix that I see is https://github.com/WyLyu/UPOsHam/issues/55. I think I'll be ready to approve once the fix for that is merged.

**Comment by Shibabrat** [comment]
@kthyng @JoshKarpel @dpsanders 
the pending PRs https://github.com/WyLyu/UPOsHam/pull/57, https://github.com/WyLyu/UPOsHam/pull/58, and https://github.com/WyLyu/UPOsHam/pull/59 are now ready for review. The major update is the package is now installable as a module and the scripts can load the system-specific functions and method-specific functions using, for example,

```
import uposham.differential_correction as diffcorr
import uposham.deleonberne_hamiltonian as deleonberne
```

As these PRs get approved, I'll merge them and run tests again.

**Comment by Shibabrat** [comment]
@kthyng @dpsanders @JoshKarpel we are ready for the next step of the review. The tests have also passed after the recent merge. How do we proceed now?

**Comment by JoshKarpel** [comment]
I'm happy with the changes, have checked off all my boxes, and recommend acceptance.

**Comment by dpsanders** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by Shibabrat** [comment]
@dpsanders I'll go through this version to check for typos and update the pdf. Do you want us to address any of the 3 remaining things in your checklist?

@JoshKarpel thanks for the approval.

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by dpsanders** [comment]
There are a few typos and English issues (mainly singular / plural) in the paper that should be checked. 

I think somewhere it needs to be made clearer exactly what is shown in the figures -- presumably contours of the Hamiltonian (which is not stated anywhere as far as I can see) and the periodic orbits computed by the code.

Other than that everything looks good and I recommend acceptance.

**Comment by kthyng** [comment]
@Shibabrat let us know when you've finished with @dpsanders comment!

**Comment by Shibabrat** [comment]
@kthyng @dpsanders we've made the corrections.

@whedon generate pdf

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by dpsanders** [comment]
Thanks, LGTM!

**Comment by Shibabrat** [comment]
@kthyng we are finished with the final changes.

**Comment by kthyng** [comment]
@Shibabrat Great! I see that the version is listed as 1.0.0. Is this still correct? I encourage you to set up a release of your code at the version you want, also: https://help.github.com/en/github/administering-a-repository/creating-releases

**Comment by kthyng** [comment]
After that, please archive your code at something like Zenodo, then report the DOI here.

**Comment by kthyng** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by kthyng** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1016/S0010-4655(97)00131-8 is OK
- 10.1063/1.166509 is OK
- 10.1103/PhysRevE.98.052214 is OK
- 10.1016/j.cnsns.2016.10.021 is OK
- 10.1103/PhysRevE.100.022204 is OK
- 10.1134/S1560354716060034 is OK
- 10.1063/1.439276 is OK
- 10.1063/1.442459 is OK
- 10.1063/1.456915 is OK
- 10.1063/1.456914 is OK

MISSING DOIs

- https://doi.org/10.1142/9789812792617_0222 may be missing for title: Dynamical systems, the three-body problem and space mission  design
- https://doi.org/10.1007/978-1-4612-3486-9 may be missing for title: Practical Numerical Algorithms for Chaotic Systems

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
I am editing your paper now and I don't know how to fix this sentence, but it is grammatically incorrect with "the dynamics":

"In two degree of freedom systems, the phase space is four dimensional and the dynamics is on the three dimensional energy surface..."

**Comment by labarba** [comment]
Hi @kthyng ... it looks fine to me, but perhaps a verb change would help. Also, add hyphens to the compound modifiers:

"In two-degree-of-freedom systems, the phase space is four dimensional and the dynamics takes place on the three-dimensional energy surface…"

**Comment by kthyng** [comment]
Also, you go between using and not using your acronym UPO frequently through the paper — seems a little odd but is ok.

I've made more small grammar changes in PR [#60](https://github.com/WyLyu/UPOsHam/pull/60) that you can merge if you agree with the changes.

**Comment by kthyng** [comment]
Here's another PR for your references: [#61](https://github.com/WyLyu/UPOsHam/pull/61)

**Comment by Shibabrat** [comment]
> @Shibabrat Great! I see that the version is listed as 1.0.0. Is this still correct? I encourage you to set up a release of your code at the version you want, also: https://help.github.com/en/github/administering-a-repository/creating-releases

I think it will be better to say version = 0.1.0. Ok, let me do those merges and changes now.

**Comment by Shibabrat** [comment]
@kthyng http://doi.org/10.5281/zenodo.3601504 which I've tagged as v0.1.0 release. Please note that at the moment doi on the readme links to the JOSS submitted version.

**Comment by kthyng** [comment]
@whedon set v0.1.0 as version

**Comment by whedon** [comment]
OK. v0.1.0 is the version.

**Comment by kthyng** [comment]
@whedon set 10.5281/zenodo.3601504 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3601504" target="_blank">10.5281/zenodo.3601504</a> is the archive.

**Comment by kthyng** [comment]
> so should I change all occurrences of degree of freedom to degree-of-freedom?

I think so, yes.

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by kthyng** [comment]
> > so should I change all occurrences of degree of freedom to degree-of-freedom?
> 
> I think so, yes.

@Shibabrat Sorry I should add that the "two-" at the beginning is important. Without that, it is two of the systems, not specifying the degree of the systems.

**Comment by kthyng** [comment]
Let me know when you're finished

**Comment by Shibabrat** [comment]
@kthyng ok. I've added the "two-" to the title and other occurrences in the paper. For consistency, I've also updated the zenodo archive and here is the doi: 10.5281/zenodo.3606676

**Comment by kthyng** [comment]
@whedon set 10.5281/zenodo.3606676 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3606676" target="_blank">10.5281/zenodo.3606676</a> is the archive.

**Comment by kthyng** [comment]
Ok I think we're good to go!!

**Comment by kthyng** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1142/9789812792617_0222 is OK
- 10.1007/978-1-4612-3486-9 is OK
- 10.1016/S0010-4655(97)00131-8 is OK
- 10.1063/1.166509 is OK
- 10.1103/PhysRevE.98.052214 is OK
- 10.1016/j.cnsns.2016.10.021 is OK
- 10.1103/PhysRevE.100.022204 is OK
- 10.1134/S1560354716060034 is OK
- 10.1063/1.439276 is OK
- 10.1063/1.442459 is OK
- 10.1063/1.456915 is OK
- 10.1063/1.456914 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1212

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1212, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kthyng** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by kthyng** [comment]
@Shibabrat I think something that you changed in your last commit altered the ability of the paper to be processed correctly. Maybe it's the line breaks, like in the title?

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1684 with the following error: 

 /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:145:in `check_fields': Paper YAML header is missing expected fields: date (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:87:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1684 with the following error: 

 /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:145:in `check_fields': Paper YAML header is missing expected fields: date (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:87:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1684 with the following error: 

 /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:145:in `check_fields': Paper YAML header is missing expected fields: date (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:87:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by Shibabrat** [comment]
@kthyng not sure what happened but I made sure the file looks markdown and pdf file looks ok on my end. Is whedon using the zenodo link to build the pdf for the paper?

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1684 with the following error: 

 /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:145:in `check_fields': Paper YAML header is missing expected fields: date (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:87:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by arfon** [comment]
@Shibabrat the errors is right there:

```
Paper YAML header is missing expected fields: date 
```

**Comment by Shibabrat** [comment]
@arfon thanks for pointing that out, the field name had a typo: 'data'

@kthyng hopefully this resolves it. going to regenerate pdf to check the compile.

**Comment by Shibabrat** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01684/joss.01684/10.21105.joss.01684.pdf)

**Comment by kthyng** [comment]
Ok great! Let's try again.

**Comment by kthyng** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1142/9789812792617_0222 is OK
- 10.1007/978-1-4612-3486-9 is OK
- 10.1016/S0010-4655(97)00131-8 is OK
- 10.1063/1.166509 is OK
- 10.1103/PhysRevE.98.052214 is OK
- 10.1016/j.cnsns.2016.10.021 is OK
- 10.1103/PhysRevE.100.022204 is OK
- 10.1134/S1560354716060034 is OK
- 10.1063/1.439276 is OK
- 10.1063/1.442459 is OK
- 10.1063/1.456915 is OK
- 10.1063/1.456914 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1217

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1217, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1217092866780073984) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1218
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01684](https://doi.org/10.21105/joss.01684)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kthyng** [comment]
Congrats to @Shibabrat and @WyLyu on your new paper!! Huge thanks to reviewers @dpsanders and @JoshKarpel for your expertise and hard work to review and improve this package. 🎉

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01684/status.svg)](https://doi.org/10.21105/joss.01684)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01684">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01684/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01684/status.svg
   :target: https://doi.org/10.21105/joss.01684
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01684/status.svg)](https://doi.org/10.21105/joss.01684)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by Shibabrat** [comment]
Thank you @JoshKarpel @dpsanders @kthyng and @labarba for the guidance, patience, and help along the way. Looking forward to contributing to the community of reproducible science and open-source software.

**Comment by dpsanders** [comment]
You're welcome. Many thanks to @kthyng and @labarba for their patience and guidance through the process.

**Comment by JoshKarpel** [comment]
Thanks all around!
